django_migration_linter/sql_analyser/postgresql.py

```diff
@@ -5,6 +5,36 @@
 from .base import BaseAnalyser, Check, CheckMode, CheckType
 
 
+def has_create_index_in_transaction(sql_statements: list[str], **kwargs) -> bool:
+    """Return if a migration opens a transaction, acquires EXCLUSIVE lock, then indexes.
+
+    Any locks that are obtained after a transaction is opened will not be released
+    until that transaction either commits or rolls back.
+
+    Accordingly, it's extremely important that a migration avoid any long-running
+    queries while exclusive locks are held.
+
+    Building an index is generally a long-running operation. A standard index
+    creation (that is, avoiding use of `CONCURRENTLY`) can be safe to run on
+    tables which are small in size and/or read-heavy. (`CREATE INDEX` only
+    prevents writes to a table; reads are allowed during index creation).
+
+    This check is a stricter version of `CREATE_INDEX` -- if a team wishes
+    to build indices nonconcurrently, it's imperative to be mindful of locks.
+    """
+    if not (sql_statements and sql_statements[0].startswith("BEGIN")):
+        return False
+
+    for i, sql in enumerate(sql_statements):
+        # If any statements acquire an exclusive lock, complain about index creation later.
+        # Nearly every single `ALTER TABLE` command requires an exclusive lock:
+        #     https://www.postgresql.org/docs/current/sql-altertable.html
+        # (Most common example is `ALTER TABLE... ADD COLUMN`, then later `CREATE INDEX`)
+        if sql.startswith("ALTER TABLE"):
+            return has_create_index(sql_statements[i + 1 :])
```

Comment:
The issue re-using this function here, is that we are not detecting some cases you are describing. For instance if we had this unit test:
```
diff --git a/tests/unit/test_sql_analyser.py b/tests/unit/test_sql_analyser.py
index 01a2bb3..7553796 100644
--- a/tests/unit/test_sql_analyser.py
+++ b/tests/unit/test_sql_analyser.py
@@ -320,6 +320,15 @@ class PostgresqlAnalyserTestCase(SqlAnalyserTestCase):
         ]
         self.assertBackwardIncompatibleSql(sql, code="CREATE_INDEX_EXCLUSIVE")

+    def test_create_concurrently_index_exclusive(self):
+        sql = [
+            "BEGIN;",
+            'ALTER TABLE "users" ADD COLUMN "email" varchar(254) NULL;',
+            'CREATE CONCURRENTLY INDEX "user_email" ON "users" ("email");',
+            "COMMIT;",
+        ]
+        self.assertBackwardIncompatibleSql(sql, code="CREATE_INDEX_EXCLUSIVE")
+
     def test_create_index_exclusive_no_lock(self):
         sql = [
             'ALTER TABLE "users" ADD COLUMN "email" varchar(254) NULL;',
```
It does not trigger an error:
```
Found 19 test(s).
System check identified no issues (0 silenced).
..F................
======================================================================
FAIL: test_create_concurrently_index_exclusive (tests.unit.test_sql_analyser.PostgresqlAnalyserTestCase.test_create_concurrently_index_exclusive)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./django-migration-linter/tests/unit/test_sql_analyser.py", line 330, in test_create_concurrently_index_exclusive
    self.assertBackwardIncompatibleSql(sql, code="CREATE_INDEX_EXCLUSIVE")
  File "./django-migration-linter/tests/unit/test_sql_analyser.py", line 30, in assertBackwardIncompatibleSql
    self.assertNotEqual(0, len(errors), "Found no errors in sql")
AssertionError: 0 == 0 : Found no errors in sql

----------------------------------------------------------------------
Ran 19 tests in 0.003s

FAILED (failures=1)
```

I guess we can modify a bit the check to also take that into account :) Since, as you said, creating an index concurrently can be a long running operation.

Reply:
Great callout and this is just oversight on my part. I'll make sure that `CREATE INDEX CONCURRENTLY` is flagged as well.